peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+
+1. ``hasattr(obj, name)``
+2. ``isinstance(obj.name, T)`` [#invalid_typevar]_
+
+The above are satisfiable at runtime by all of the following:
+
+1. an object with an attribute ``name: T``,
+2. a class with a class variable ``name: ClassVar[T]``, [#invalid_typevar]_
+3. an instance of the class above,
+4. an object with a ``@property`` ``def name(self) -> T``,
+5. an object with a custom descriptor, such as :func:`functools.cached_property`.
+
+Note that the attribute being marked ``Final`` or the property defining a setter
+do not impact this.
+
+The most common practice is to define such a protocol with a ``@property``::
+
+    class HasName(Protocol):
+        @property
+        def name(self) -> T: ...
+
+Type checkers special-case this definition, such that objects with plain attributes
+are assignable to the type. However, instances with class variables and descriptors
+other than ``property`` are rejected.
+
+Covering the extra possibilities induces a great amount of boilerplate, involving
+creation of an abstract descriptor protocol, possibly also accounting for
+class and instance level overloads.
+Worse yet, all of that is multiplied for each additional read-only attribute.
+
+
+Rationale
+=========
+
+These problems can be resolved by an attribute-level :external+typing:term:`type qualifier`.
+``ReadOnly`` has been chosen for this role, as its name conveys the intent well,
+and the newly proposed changes complement its semantics defined in the :pep:`705`.
+
+A class with a read-only instance attribute can be now defined as such::
+
+    from typing import ReadOnly
+
+
+    class Member:
+        def __init__(self, id: int) -> None:
+            self.id: ReadOnly = id
+
+...and a protocol described in :ref:`protocols` is now just::
+
+    from typing import Protocol, ReadOnly
+
+
+    class HasName(Protocol):
+        name: ReadOnly[str]
+
+
+    def greet(obj: HasName, /) -> str:
+        return f"Hello, {obj.name}!"
+
+* A subclass of ``Member`` can redefine ``.id`` as a writable attribute or a
+  :term:`descriptor`. It can also :external+typing:term:`narrow` the type.
+* The ``HasName`` protocol can be implemented by any mechanism allowing for ``.name`` access.
+* The ``greet`` function can now accept a wide variety of compatible objects,
+  while being explicit about no modifications being done to the input.
+
+
+Specification
+=============
+
+The :external+py3.13:data:`typing.ReadOnly` :external+typing:term:`type qualifier`
+becomes a valid annotation for :term:`attributes <attribute>` of classes and protocols.
+It is used to indicate that an attribute should not be reassigned or ``del``\ eted.
+
+The deletability rule should be extended to ``Final`` as well, as it is currently
+not specified.
+
+Akin to ``Final``, read-only attributes do not influence the mutability of
+the assigned object. Immutable ABCs and containers may be used in combination with
+``ReadOnly`` to prevent mutation of such values.
```

Comment:
This comes from https://typing.readthedocs.io/en/latest/spec/qualifiers.html#semantics-and-examples
> Note that declaring a name as final only guarantees that the name will not be re-bound to another value, but does not make the value immutable. Immutable ABCs and containers may be used in combination with Final to prevent mutating such values: [...]

Unfortunately it's fairly deep into that section :/

> For example, given a `ReadOnly` member `names` of an object `x`, `x.names.append("Pat")`
is not an error.

There's an example of a mutable read-only attribute later in the `Initialization` section, I suppose I could float it closer to this paragraph